### ISSUE TITLE
Ragged Arrays of Complex User Types

### ISSUE THREAD

**Issue by rosenbrockc** [rosenbrockc/fortpy#59]
It turns out there is another, _really_ nasty use case for the automated unit testing that isn't handled yet. Suppose I have a derived type such as:

```fortran
type nasty
  integer :: A(:,:)
  logical :: B(:)
  complex :: C(:,:,:)
end type nasty
```

Next, suppose that the subroutine I am unit testing uses a 3D array of these guys:

```fortran
subroutine uses_nasty(narray, oarray)
  type(nasty), intent(in) :: narray(:,:,:)
  type(nasty), intent(out) :: oarray(:,:)
end subroutine uses_nasty
```

To initialize the input parameter, we would need a bunch of files storing the arrays. For example `A.in.{}`, `B.in.{}` and `C.in.{}`. Since `C` is already 3D, there is no hope of doing this using a ragged array setup. Instead, we could use `<part>` tags to nest the assignment of the values to these members of the derived type `nasty`.

```xml
<assignment name="narray" value="p1" allocate="10,5,2">
  <value identifier="A" file="A.in.{}.*" folder="./tests" member="A" suffix="$1.$2.$3" rename="A.in" />
  <value identifier="B" file="B.in.{}.*" folder="./tests" member="B" suffix="$1.$2.$3" rename="B.in" />
  <value identifier="C" file="C.in.{}.*" folder="./tests" member="C" suffix="$1.$2.$3" rename="C.in" />
  <part identifier="p1">
    <part identifier="p2">
      <part identifier="p3" value="A, B, C" />
    </part>
  </part>
</assignment>
```

The new _suffix_ attribute specifies the text (where loop variable values are substituted for the `$i` values) to replace `*` with in the file name when searching for the source to load a variable's value from. When the driver is generated, a call to `fpy_period_join_indices` will generate the suffix and it will be spliced in by python in the right place using `filename.replace("*", "// varname_fpy_suffix //")`. The wildcard in the filename will still ensure that the existing routines will copy all the necessary files to the execution directory for a given test case. The above XML says to allocate `narray(10,5,2)` and then create a loop system:

```fortran
do narray_fpy_1=1, size(narray, 1)
  do narray_fpy_2=1, size(narray, 2)
    do narray_fpy_3=1, size(narray, 3)
      call fpy_period_join_indices(narray_fpy_suffix, (/ narray_fpy_1, narray_fpy_2, narray_fpy_3 /), 3)
      call fpy_read(narray(narray_fpy_1, narray_fpy_2, narray_fpy_3)%A, "A.in." // narray_fpy_suffix)
      call fpy_read(narray(narray_fpy_1, narray_fpy_2, narray_fpy_3)%B, "B.in." // narray_fpy_suffix)
      call fpy_read(narray(narray_fpy_1, narray_fpy_2, narray_fpy_3)%C, "C.in." // narray_fpy_suffix)
    end do
  end do
end do
```

Things to do:

1. Add support for the _suffix_ attribute to the `<value>` tags. Also the _member_ attribute is still awaiting implementation.
2. Replace the `*` with the appropriate suffix code (which includes initializing a new suffix variable and calling `fpy_period_join_indices`) in all calls to `fpy_read`.
3. Allow a comma-separated list of `<value>` identifiers to be specified for `<part>` _and_ `<assignment>` tags. The order they appear in the comma-separated list determines the order in which they are executed. Thus, a user could assign values to `A, B, C` above using those `<value>` tags and then still call an embedded function (for example).

**Comment by rosenbrockc** [comment]
The above fix for _reading_ arbitrarily structured user-derived types from multiple files is implemented in Revision 1.6.1 and can handle arbitrary arrays of these types (i.e. the input required for the `uses_nasty` subroutine above can be specified easily as described). I also updated the wiki with the new attributes. As part of this fix, members can have their values set now as well with the _member_ attribute.

We now turn attention to the second part of the unit test problem: we need to save the output of the arrays of user-defined types to files and then check them. Obviously, for large arrays of complex types, it is not possible to specify hundreds of `<target>` and `<output>` tags to test everything (well I suppose it is possible, but the developer will jump off a building before they duplicate so many similar lines of code). It would be nice if we could automatically save user-derived types to file in a logical way. How this would work:

- Instead of working with a _file_ directly, we would allow the _folder_ attribute_ to specify the contents for the variable (which is an array of user-derived types).
- We would introduce attribute _autoclass_ which tells fortpy to generate files for the entire contents of the variable. It would save a file such as `varname.$1.$2.$3.member` for each user-derived variable and all of its members. (in this example it is a 3D array of them and `$i` is the loop variable for iterating over them).
- The `<output>` tag will also have the _autoclass_ attribute, which would modify the meaning of the _folder_ attribute to point to the contents of the variable. Each file in that folder should follow the naming convention specified above **with lower-case names for variables and members**.

Here is an example:

```xml
<output folder="../tests/derivative_structure_generator/organize_rot_perm_lists/rdRPList"
             autoclass="true" identifier="rdRPList.out" />
<target name="rdRPList" autoclass="true" compareto="rdRPList.out" varname="rdRPList" />
```

The _varname_ attribute of the `<target>` tag now means the _folder_ that will be created to house all the contents of the variable. The `<output>` tag no longer treats the _template_ tag since fortpy will auto-add the template needed to the first line of each output file and use those by default. When the comparison runs, it will run for _every_ file present in each folder. If any files are missing in either folder, it will whine appropriately.

Things to do:

1. Add the _autoclass_ attribute to `<target>` and `<output>` tags.
2. Create a method to auto-generate the output folder with pysave; we need to check how Fortran handles relative paths in the `open` built-in.
3. For the comparison, write a wrapper method for the existing `<output>` comparison routines; call it for each file that matches in the model and test-output folders. Print warnings/errors for mismatches between the folders.

**Comment by rosenbrockc** [comment]
The _autoclass_ attribute is now implemented starting Revision 1.6.1 including comparisons. It is not extensively tested yet, so expect revisions in future minor releases.

There is also another new attribute _actolerance_ that sets the failure tolerance for any individual file compared within the auto-class framework.
